File: khan_mini/essaylib/scoring.py
import numpy
from numpy import matrix
from numpy import array
from numpy import linalg


def parknewman(A, alpha=0.2):
    N = A.shape[0]

    I = numpy.identity(N)
    kout = (matrix(numpy.ones(N))* A).T
    kin = A * matrix(numpy.ones(N)).T
    w = (I - alpha* A.T)
    w = w.I*kout

    l = (I - alpha* A)
    l = l.I*kin

    return w-l
    

def colley(A):
    N = A.shape[0]
    kout = (matrix(numpy.ones(N))* A).T
    kin = A * matrix(numpy.ones(N)).T

    b = 1+(kout - kin)/2
    C  = matrix(numpy.identity(N))
    for i in range(N):
        for j in range(N):
            if(i==j): 
                C[i,j] = ((kout[i] + kin[i]) + 2)
            else:
                C[i,j] = -(A[i,j] + A[j,i])


    return linalg.solve(C, b)

def standardize(r):
    r = (r - numpy.mean(r))/ numpy.std(r)
    return r

 

def test():
    A = numpy.matrix([[0,1,1,0,0,0],
                      [0,0,0,0,0,0],
                      [1,1,0,0,1,0],
                      [0,0,0,0,1,1],
                      [0,0,0,1,0,1],
                      [0,0,0,1,0,0]], dtype=numpy.float) 
    p = parknewman(A)
    c = colley(A)
    p = standardize(p)
    c = standardize(c)
    print "Park Newman\tColley"
    print "\n".join(["%s\t%s" % (p.item(i), c.item(i)) for i in range(A.shape[0])])


if __name__ == "__main__":
    test()

File: PeerMarker/app/models/scoring.rb
require 'matrix'
def parknewman(matrixA, alpha=0.2)
    n = matrixA.row_size

    i = Matrix.unit(n)
    ones = Matrix.[]([1]*n) 
    kout = (ones* matrixA).transpose
    kin = matrixA * (ones.transpose)
    w = (i - alpha* matrixA.transpose)
    w = w.inverse*kout

    l = (i - alpha* matrixA)
    l = l.inverse*kin

    return w-l
end    

def colley(matrixA)
    n = matrixA.row_size
    ones = Matrix.[]([1]*n)
    kout = (ones * matrixA).transpose
    kin = matrixA * (ones.transpose)

    b = ones.transpose + (kout - kin)*0.5

    c  = Matrix.unit(n)
    
    n.times do |i|
        n.times do |j|
            if(i==j) then
                c[i,j] = ((kout[i,0] + kin[i,0]) + 2)
            else
                c[i,j] = -(matrixA[i,j] + matrixA[j,i])
            end
        end
    end            

    return b.transpose/c
end

def standardize(r)
    return r.round 2
end    

class Matrix
  def []=(i, j, x)
    @rows[i][j] = x
  end
end 

def test()
    matrixA = Matrix[[0,1,1,0,0,0],[0,0,0,0,0,0], [1,1,0,0,1,0],[0,0,0,0,1,1],[0,0,0,1,0,1],[0,0,0,1,0,0]]
    p = parknewman(matrixA)
    c = colley(matrixA)
    p = standardize(p)
    c = standardize(c.transpose)
    n = matrixA.row_size
    puts "Park Newman\tColley"
    puts n.times.map {|i| "#{p[i,0]}\t#{c[i,0]}"}
end

if __FILE__ == $PROGRAM_NAME
   test()    
end
